File: chisel4ml/scala/main/Circuit.scala
/*
 * Copyright 2022 Computer Systems Department, Jozef Stefan Insitute
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package chisel4ml

import chisel3._
import chisel4ml._
import chisel4ml.implicits._
//import chiseltest._
//import chiseltest.simulator.{WriteFstAnnotation, WriteVcdAnnotation}
//import firrtl2.AnnotationSeq
//import firrtl2.options.TargetDirAnnotation
//import firrtl2.transforms.NoCircuitDedupAnnotation
import svsim._
import chisel3.simulator._
import chisel4ml.simulator._
import chisel4ml.simulator.simulatorSettings._
import lbir.QTensor
import memories.MemoryGenerator
import org.slf4j.LoggerFactory

import java.util.concurrent.{CountDownLatch, LinkedBlockingQueue, TimeUnit}
import scala.util.control.Breaks._

/** Contains the generated hardware module and provides a simulation interface
  *
  * @param dutGen
  *   The generated hardware container
  * @param outputStencil
  *   The expected output QTensor shape and datatype
  * @param directory
  *   The directory that holds the generated FIRRTL/Verilog files
  * @param useVerilator
  *   Use verilator for simulation?
  * @param genWaveform
  *   Generate waveform during RTL simulation?
  * @param waveformType
  *   The type of waveform to generate during RTL simulation. Either 'fst' or 'vcd'.
  */
class Circuit[+T <: Module with HasAXIStream](
  dutGen:        => T,
  outputStencil: QTensor,
  directory:     os.Path,
  useVerilator:  Boolean,
  genWaveform:   Boolean,
  waveformType:  String = "fst")
    extends Runnable {
  case class ValidQTensor(qtensor: QTensor, valid: Boolean)
  case class TimedQTensor(qtensor: QTensor, consumedCycles: Int)
  val logger = LoggerFactory.getLogger(classOf[Circuit[T]])
  val inQueue = new LinkedBlockingQueue[ValidQTensor]()
  val outQueue = new LinkedBlockingQueue[TimedQTensor]()
  val isGenerated = new CountDownLatch(1)
  val isStoped = new CountDownLatch(1)
  val relativeDirectory = directory.relativeTo(os.pwd).toString()

  var simSettings: Seq[SimulatorSettings] = Seq(
    SaveWorkdirFile(relativeDirectory)
  )
  if (genWaveform) {
    if (waveformType.toLowerCase == "vcd") {
      simSettings = simSettings :+ VcdTrace
    } else {
      simSettings = simSettings :+ VcdTrace
    }

  }
  if (useVerilator) simSettings = simSettings

  def stopSimulation(): Unit = {
    inQueue.put(ValidQTensor(QTensor(), false))
    isStoped.await(5, TimeUnit.SECONDS)
  }

  def run(): Unit = {
    logger.info(s"Used settings for generated circuit are: ${simSettings.map(_.toString)}.")
    MemoryGenerator.setGenDir(directory)
    ParametricSimulator.simulate(
      dutGen,
      simSettings
    )(this.simulate(_))
    isStoped.countDown()
  }

  /** Simulates a circuit using an input and output queue of QTensors.
    *
    * @param dut
    *   the hardware module to simulate.
    */
  private[this] def simulate(dut: T): Unit = {
    isGenerated.countDown()
    logger.info(s"Generated circuit in directory: ${directory}.")
    dut.clock.step(200) // TODO
  }

  /** The user interface to the QTensor based simulation
    *
    * @param x
    *   A sequence of input QTensors to simulate
    * @return
    *   Resulting QTensors and the number of cycles used to simulate it.
    */
  def sim(x: Seq[QTensor]): (Seq[QTensor], Int) = {
    var result:         Seq[QTensor] = Seq()
    var consumedCycles: Int = 0
    for (qtensor <- x) {
      inQueue.put(ValidQTensor(qtensor, true))
    }
    for (_ <- x) {
      val out = outQueue.take()
      result = result :+ out.qtensor // .take() is a blocking call
      consumedCycles = consumedCycles + out.consumedCycles
    }
    (result, consumedCycles)
  }
}

File: chisel4ml/circuit.py
# Copyright 2022 Computer Systems Department, Jozef Stefan Insitute
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#  https://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import logging
import os
import shutil
from pathlib import Path

import numpy as np

import chisel4ml.lbir.services_pb2 as services
from chisel4ml import transforms
from chisel4ml.chisel4ml_server import Chisel4mlServer
from chisel4ml.chisel4ml_server import connect_to_server
from chisel4ml.lbir.qtensor_pb2 import QTensor

log = logging.getLogger(__name__)


class Circuit:
    """Container class for the generated circuits. Objects of type Circuit hold a
    reference to the circuit implemnentation in the chisel4ml server memory. It also
    provides a python interface to that server via gRPC (via __call__ or predict).
    """

    def __init__(
        self,
        circuit_id: int,
        input_quantizer,
        input_qtensor: QTensor,
        lbir_model,
        server: Chisel4mlServer = None,
    ):
        assert circuit_id >= 0, (
            "Invalid circuitId provided. This parameter should be positive, but is"
            f" {circuit_id}."
        )
        self.circuit_id = circuit_id
        self.input_quantizer = input_quantizer
        self.input_qtensor = input_qtensor
        if server is None:
            self._server = connect_to_server()
        else:
            self._server = server
        self.lbir_model = lbir_model
        self.consumed_cycles = None

    def __call__(self, np_arr, sim_timeout_sec=200):
        "Simulate the circuit, timeout in seconds."
        qtensors = transforms.numpy_transforms.numpy_to_qtensor(
            np_arr, self.input_quantizer, self.input_qtensor
        )
        run_sim_params = services.RunSimulationParams(
            circuit_id=self.circuit_id, inputs=qtensors
        )
        run_sim_return = self._server.send_grpc_msg(
            run_sim_params, timeout=sim_timeout_sec
        )
        self.consumed_cycles = run_sim_return.consumed_cycles
        results = []
        for res in run_sim_return.values:
            results.append(np.array(res.values).reshape(res.shape))
        return np.concatenate(results, axis=0)

    def predict(self, np_arr):
        return self.__call__(np_arr)

    def delete_from_server(self):
        delete_circuit_params = services.DeleteCircuitParams(circuit_id=self.circuit_id)
        delete_circuit_return = self._server.send_grpc_msg(
            delete_circuit_params, timeout=10
        )
        log.info(delete_circuit_return.msg)
        return delete_circuit_return.success

    def package(self, directory=None):
        if directory is None:
            raise ValueError("Directory parameter missing.")
        temp_dir = self._server.temp_dir
        temp_circuit_dir = os.path.join(temp_dir, f"circuit{self.circuit_id}")

        def get_files(extensions, directory):
            all_files = []
            for ext in extensions:
                all_files.extend(Path(directory).glob(ext))
            return all_files

        files = get_files(("*.sv", "*.bin", "*.hex"), directory=temp_circuit_dir)
        os.makedirs(Path(directory).absolute(), exist_ok=True)
        for file in files:
            dest_file = os.path.join(directory, os.path.split(file)[1])
            shutil.copyfile(file, dest_file)
